File: wallet/util/jwe_util.py
# Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# -*- coding: utf-8 -*-
# @Date     : 2020/06/12 17:43
# @Software : PyCharm
# @version  ：Python 3.6.8
# @File     : jwe_util.py

from wallet.util.jwe_header import JweHeader
from wallet.util import aes_util
from wallet.util import rsa
import base64
import secrets


def get_header():
    jwe_header = JweHeader()
    jwe_header.set_alg("RSA-OAEP")
    jwe_header.set_enc("A128GCM")
    jwe_header.set_kid("1")
    jwe_header.set_zip("gzip")
    return jwe_header


def get_encode_header(jwe_header):
    header_data = "alg=%s, enc=%s, kid=%s, zip=%s" % (
        jwe_header.get_alg(), jwe_header.get_enc(), jwe_header.get_kid(), jwe_header.get_zip())

    header_str_byte = bytes(header_data, encoding='utf-8')
    jwe_header_encode = base64.urlsafe_b64encode(header_str_byte)
    return jwe_header_encode.decode('utf-8')


def get_encrypted_key(session_key, session_key_public_key):
    encrypt_data = rsa.encrypt_by_rsa(session_key, session_key_public_key)
    encrypt_data_encode = base64.urlsafe_b64encode(encrypt_data)
    print("Encrypted sessionKey: ", encrypt_data_encode)
    return encrypt_data_encode.decode('utf-8')


def get_cipher_text(pay_load, session_key, iv_hex):
    payload_encrypt = aes_util.encrypt_by_gcm(pay_load, session_key, iv_hex)
    print("Encrypted payload Hex String: ", payload_encrypt)
    payload_encrypt_compress_byte = aes_util.compress_data(payload_encrypt)
    cipher_text_encode = base64.urlsafe_b64encode(payload_encrypt_compress_byte)
    return cipher_text_encode.decode('utf-8')


def get_signature(jwe_private_key, session_key, pay_load, jwe_header_encode, iv_hex_encode):
    sign_content = jwe_header_encode + "." + session_key + "." + iv_hex_encode + "." + pay_load
    print("Content to be signed: ", sign_content)
    sign_data_byte = rsa.sign_by_private_key(sign_content, jwe_private_key)
    sign_data = sign_data_byte.decode('utf-8')
    return sign_data


def generate_jwe(jwe_private_key, pay_load):
    # Part 1: JWE Header.
    # This header contains information about encryption and compression algorithms. It's a constant String.

    print("Part 1:")
    jwe_header = get_header()
    jwe_header_encode = get_encode_header(jwe_header)
    print("Encoded header: ", jwe_header_encode)

    # Part 2: JWE Encrypted Key
    # Generate a 16-byte session key to encrypt payload data. Then convert it to a Hex String.
    print("Part 2:")
    session_key = secrets.token_hex(16)
    print("session key: ", session_key)
    session_key_public_key = "MIIBojANBgkqhkiG9w0BAQEFAAOCAY8AMIIBigKCAYEAgBJB4usbO33Xg5vhJqfHJsMZj44f7rxpjRuPhGy37bUBjSLXN+dS6HpxnZwSVJCtmiydjl3Inq3Mzu4SCGxfb9RIjqRRfHA7ab5p3JnJVQfTEHMHy8XcABl6EPYIJMh26kztPOKU2Mkn6yhRaCurhVUD3n9bD8omiNrR4rg442AJlNamA7vgKs65AoqBuU4NBkGHg0VWWpEHCUx/xyX6hIwqc1aD7P2f62ZHsKpNZBOek/riWhaVx3dTAa9ZS+Av3IGLOZiplhYIow9f8dlWyqs8nff9FZoJO03QhXLvOORT+lPAkW6gFzaoeMaGb40HakkZn3uvlAEKrKrtR0rZEok+N1hnboaAu8oaKK0rF1W6iNrXcFrO0rcrCsFTVF8qCa/1dFmIXwUd2M6cUzT9W0YkNyb6ZBbwEhjwBL4DNW4JfeF2Dzj0eZYlSuYV7e7e1e+XEO8lwPLAiy4bEFAWCaeuDVIhbIoBaU6xHNVQoyzct98gaOYxE4mVDqAUVmhfAgMBAAE="

    session_key_public_key_pem = "-----BEGIN PUBLIC KEY-----\n%s\n-----END PUBLIC KEY-----\n" % session_key_public_key

    encrypted_key_encode = get_encrypted_key(session_key, session_key_public_key_pem)
    print("Encrypted sessionKey: ", encrypted_key_encode)

    # Part 3: JWE IV
    # Generate a 12-byte iv. Then convert it to a Hex String, and then do base64 encoding to the Hex String.
    print("Part 3:")
    iv_hex = secrets.token_hex(12)
    iv_hex_byte = bytes(iv_hex, encoding='utf-8')
    iv_hex_encode_byte = base64.urlsafe_b64encode(iv_hex_byte)
    iv_hex_encode = iv_hex_encode_byte.decode('utf-8')
    print("Encoded iv: ", iv_hex_encode)

    # Part 4: JWE Cipher Text
    # Encrypt the payload with sessionKey and iv using AES/GCM/NoPadding algorithm. Encode the cipher text into a
    # Hex String. Then do gzip compression and base64 encoding to the Hex String.
    print("Part 4:")
    print("pay_load: " + pay_load)
    cipher_text_encode = get_cipher_text(pay_load, session_key, iv_hex)
    print("Compressed and encoded cipher text: ", cipher_text_encode)

    # Part 5: JWE Signature
    # Use your own private key to sign the content with SHA256withRSA, then do base64 encoding to it.
    # HMS wallet server will use the public key you provided on AGC to verify signatures.
    print("Part 5:")
    signature = get_signature(jwe_private_key, session_key, pay_load, jwe_header_encode, iv_hex_encode)
    print("signature: ", signature)

    # Combine all five parts together to form a valid JWE.
    jwe_data = "%s.%s.%s.%s.%s" % (
        jwe_header_encode, encrypted_key_encode, iv_hex_encode, cipher_text_encode, signature)
    return jwe_data

File: src/main/java/com/huawei/wallet/util/JweUtil.java
/*
 * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package com.huawei.wallet.util;

import org.apache.commons.codec.binary.Base64;
import org.apache.commons.codec.binary.Hex;

import java.io.ByteArrayOutputStream;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.util.HashMap;
import java.util.Map;
import java.util.zip.GZIPOutputStream;

/**
 * JWE utility class.
 *
 * @since 2020-03-02
 */
public class JweUtil {

    /**
     * Generate a JSON Web Encryption (JWE).
     *
     * @param jwePrivateKey private key used to sign JWE content.
     * @param payload information plain text. It can be a list of instance IDs or an instance.
     * @return return a map containing a content string and a signature string.
     */
    public static String generateJwe(String jwePrivateKey, String payload) {
        // Part 1: JWE Header.
        // This header contains information about encryption and compression algorithms. It's a constant String.
        System.out.println("Part 1:");
        Map<String, String> jweHeader = getHeader();
        String jweHeaderEncode = getEncodeHeader(jweHeader);
        System.out.println("Encoded header: " + jweHeaderEncode);

        // Part 2: JWE Encrypted Key
        // Generate a 16-byte session key to encrypt payload data. Then convert it to a Hex String.
        System.out.println("Part 2:");
        String sessionKey = generateSecureRandomFactor(16);
        // Encrypt the session key Hex String with Huawei's fixed sessionKeyPublicKey using
        // RSA/NONE/OAEPwithSHA-256andMGF1Padding algorithm, and then do base64 encoding to it.
        System.out.println("sessionKey: " + sessionKey);
        String sessionKeyPublicKey =
            "MIIBojANBgkqhkiG9w0BAQEFAAOCAY8AMIIBigKCAYEAgBJB4usbO33Xg5vhJqfHJsMZj44f7rxpjRuPhGy37bUBjSLXN+dS6HpxnZwSVJCtmiydjl3Inq3Mzu4SCGxfb9RIjqRRfHA7ab5p3JnJVQfTEHMHy8XcABl6EPYIJMh26kztPOKU2Mkn6yhRaCurhVUD3n9bD8omiNrR4rg442AJlNamA7vgKs65AoqBuU4NBkGHg0VWWpEHCUx/xyX6hIwqc1aD7P2f62ZHsKpNZBOek/riWhaVx3dTAa9ZS+Av3IGLOZiplhYIow9f8dlWyqs8nff9FZoJO03QhXLvOORT+lPAkW6gFzaoeMaGb40HakkZn3uvlAEKrKrtR0rZEok+N1hnboaAu8oaKK0rF1W6iNrXcFrO0rcrCsFTVF8qCa/1dFmIXwUd2M6cUzT9W0YkNyb6ZBbwEhjwBL4DNW4JfeF2Dzj0eZYlSuYV7e7e1e+XEO8lwPLAiy4bEFAWCaeuDVIhbIoBaU6xHNVQoyzct98gaOYxE4mVDqAUVmhfAgMBAAE=";
        String encryptedKeyEncode = getEncryptedKey(sessionKey, sessionKeyPublicKey);
        System.out.println("Encrypted sessionKey: " + encryptedKeyEncode);

        // Part 3: JWE IV
        // Generate a 12-byte iv. Then convert it to a Hex String, and then do base64 encoding to the Hex String.
        System.out.println("Part 3:");
        byte[] iv = AesUtil.getIvByte(12);
        String ivHexStr = new String(Hex.encodeHex(iv, false));
        String ivEncode = Base64.encodeBase64URLSafeString(ivHexStr.getBytes(StandardCharsets.UTF_8));
        System.out.println("Encoded iv: " + ivEncode);

        // Part 4: JWE Cipher Text
        // Encrypt the payload with sessionKey and iv using AES/GCM/NoPadding algorithm. Encode the cipher text into a
        // Hex String. Then do gzip compression and base64 encoding to the Hex String.
        System.out.println("Part 4:");
        System.out.println("payload: " + payload);
        String cipherTextEncode = getCipherText(payload, sessionKey, iv);
        System.out.println("Compressed and encoded cipher text: " + cipherTextEncode);

        // Part 5: JWE Signature
        // Use your own private key to sign the content with SHA256withRSA, then do base64 encoding to it.
        // HMS wallet server will use the public key you provided on AGC to verify signatures.
        System.out.println("Part 5:");
        String signature = getSignature(jwePrivateKey, sessionKey, payload, jweHeaderEncode, ivEncode);
        System.out.println("signature: " + signature + "\n");

        // Combine all five parts together to form a valid JWE.
        StringBuilder stringBuilder = new StringBuilder().append(jweHeaderEncode)
            .append(".")
            .append(encryptedKeyEncode)
            .append(".")
            .append(ivEncode)
            .append(".")
            .append(cipherTextEncode)
            .append(".")
            .append(signature);
        return stringBuilder.toString();
    }

    private static Map<String, String> getHeader() {
        Map<String, String> jweHeader = new HashMap<>();
        jweHeader.put("alg", "RSA-OAEP");
        jweHeader.put("enc", "A128GCM");
        jweHeader.put("kid", "1");
        jweHeader.put("zip", "gzip");
        return jweHeader;
    }

    private static String getEncodeHeader(Map<String, String> jweHeader) {
        StringBuilder stringBuilder = new StringBuilder();
        String headerStr = stringBuilder.append("alg=")
            .append(jweHeader.get("alg"))
            .append(", enc=")
            .append(jweHeader.get("enc"))
            .append(", kid=")
            .append(jweHeader.get("kid"))
            .append(", zip=")
            .append(jweHeader.get("zip"))
            .toString();
        System.out.println("Header before encoding: " + headerStr);
        // Do base64 encoding.
        return Base64.encodeBase64URLSafeString(headerStr.getBytes(StandardCharsets.UTF_8));
    }

    private static String getEncryptedKey(String sessionKey, String sessionKeyPublicKey) {
        try {
            String encryptedSessionKey = RsaUtil.encrypt(sessionKey.getBytes(StandardCharsets.UTF_8),
                sessionKeyPublicKey, "RSA/NONE/OAEPwithSHA-256andMGF1Padding");
            return Base64.encodeBase64URLSafeString(encryptedSessionKey.getBytes(StandardCharsets.UTF_8));
        } catch (Exception e) {
            throw new IllegalArgumentException("Encrypt session key failed. Error: " + e.getMessage());
        }
    }

    private static String getCipherText(String dataJson, String sessionKey, byte[] iv) {
        String payloadEncrypt = AesUtil.encryptByGcm(dataJson, sessionKey, iv);
        System.out.println("Encrypted payload Hex String: " + payloadEncrypt);
        byte[] payLoadEncryptCompressByte = compress(payloadEncrypt.getBytes(StandardCharsets.UTF_8));
        return Base64.encodeBase64URLSafeString(payLoadEncryptCompressByte);
    }

    private static String getSignature(String jweSignPrivateKey, String sessionKey, String payLoadJson,
        String jweHeaderEncode, String ivEncode) {
        StringBuilder stringBuilder = new StringBuilder();
        String signContent = stringBuilder.append(jweHeaderEncode)
            .append(".")
            .append(sessionKey)
            .append(".")
            .append(ivEncode)
            .append(".")
            .append(payLoadJson)
            .toString();
        System.out.println("Content to be signed: " + signContent);
        return RsaUtil.sign(signContent, jweSignPrivateKey);
    }

    /**
     * gzip Compress
     *
     * @param originalBytes Data to be compressed
     * @return Compressed data
     */
    public static byte[] compress(byte[] originalBytes) {
        if (originalBytes == null || originalBytes.length == 0) {
            return null;
        }
        try (ByteArrayOutputStream out = new ByteArrayOutputStream();
            GZIPOutputStream gzip = new GZIPOutputStream(out)) {
            gzip.write(originalBytes);
            gzip.finish();
            return out.toByteArray();
        } catch (Exception e) {
            System.out.println(e.getMessage());
            return null;
        }
    }

    /**
     * Generate random hex string.
     *
     * @param size string length.
     * @return the generated random hex string.
     */
    private static String generateSecureRandomFactor(int size) {
        SecureRandom random = new SecureRandom();
        byte[] bytes = new byte[size];
        random.nextBytes(bytes);
        return org.bouncycastle.util.encoders.Hex.toHexString(bytes);
    }
}
